### ISSUE TITLE
DeepSpeed + LoRA is repaired in sft.py only: the other trainer wrappers still hit the empty no-decay group

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#359]
#336 fixed the empty-parameter-group failure (with LoRA, HF's no-decay optimizer group comes out empty, DeepSpeed drops it, and torch's scheduler then hits a 1-vs-2 strict-`zip` mismatch on `base_lrs`) — but only in `trainer/sft.py`. Every other trainer wrapper builds its own optimizer/scheduler and still fails the same way under `--deepspeed` with LoRA.

Recorded as a Known Limitation in the v0.73.0 release notes rather than left implicit.

## Fix path

`prune_empty_param_groups` already exists in `utils/deepspeed.py` and does the right thing (drops empty groups in place before `create_scheduler` runs; refuses when *every* group is empty so "nothing is trainable" stays a clear error). The work is threading it through the remaining wrappers the way `sft.py` does, and deciding where it belongs so a nineteenth wrapper cannot forget it — the `device_map` fix in the same release moved from a hand-written trainer list to a scanner over `soup_cli/trainer/` for exactly this reason.

Two related gaps from the same investigation, in scope or trackable separately:

- a user-supplied `--deepspeed my.json` is passed through unresolved, so none of the preset rewrites (bf16 vs fp16 quantised weights, `zero_hpz_partition_size` derived from `WORLD_SIZE`) apply to it;
- ZeRO++ hierarchical partitioning is unit-tested but has never been exercised across two nodes.

## Acceptance criteria

- [ ] Every trainer wrapper that builds an optimizer goes through the shared guard, enforced by a test that **scans** `soup_cli/trainer/` rather than listing names
- [ ] A control proving the scanner can fail
- [ ] At least one non-SFT wrapper verified end-to-end on real multi-GPU hardware (`benchmarks/gate-h100-validation.md` STEP 20 is the template: LoRA + DeepSpeed arms red before, green after, with the full-fine-tuning control untouched)
- [ ] Decide and document whether `--deepspeed <file>` should be resolved or deliberately left as an escape hatch

**Comment by MakazhanAlpamys** [comment]
Taking this one, with one criterion I cannot close myself — stating that up front per the claim rule.

**What I can deliver: criteria 1, 2 and 4.** Measured the current state first: 19 modules under `src/soup_cli/trainer/` accept `deepspeed_config`, and `attach_empty_param_group_guard` is called in exactly one of them (`sft.py:658`). So 18 wrappers still build a scheduler over two parameter groups that DeepSpeed will hand back as one.

**What I cannot: criterion 3.** I have no multi-GPU hardware, so I cannot produce the `benchmarks/gate-h100-validation.md` STEP 20 equivalent — LoRA + DeepSpeed arms red before and green after on a non-SFT wrapper, with the full-fine-tuning control untouched. I would rather leave that box unticked and say so than have the issue closed on a green unit suite that structurally cannot observe the failure it is about: the defect appears at `lr_scheduler.step()` under a real DeepSpeed engine, and no CPU test reaches that.

**On where the guard belongs.** The fix path notes the `device_map` fix moved from a hand-written trainer list to a scanner, and I am mirroring that rather than inventing a second shape. There is no shared construction path to hook — the wrappers have no common base class (only `StreamingSetupMixin` on four, and `SFTTrainerWrapper` subclassed by bitnet/tts) — and the guard must run at optimizer-construction time, before the scheduler exists, so it cannot be a callback. That leaves per-wrapper attachment with the coverage enforced by a scan over `soup_cli/trainer/*.py`, exactly as `tests/test_device_map_distributed.py::TestNoTrainerStillHardcodesAuto` does. Its history is the argument: the first version of that class parametrized over six hand-written names and passed while nine more sites still carried the defect.

The scanner gets its own control (criterion 2) — a test that the pattern matches the unguarded shape and does not match the guarded one, because a scanner nobody has watched fail is indistinguishable from a broken one.

**On criterion 4** (`--deepspeed <file>` resolved or left as an escape hatch): I will read what the preset rewrites actually do — bf16-vs-fp16 for quantised weights, `zero_hpz_partition_size` from `WORLD_SIZE` — and document the decision either way in the same PR. My prior is that silently not applying them is the worse of the two, since the failure is a slow one, but I will say which after reading rather than before.

The two-node ZeRO++ gap is out of scope here; it needs the same hardware criterion 3 does.

**Comment by MakazhanAlpamys** [comment]
PR up: #484 — criteria 1, 2 and 4. Criterion 3 stays unticked and I have not closed it: no CPU test in that PR can observe a failure that only appears under a live DeepSpeed engine, so the multi-GPU end-to-end still needs someone with the box. Details in the PR.

**Comment by Srinivasan8888** [comment]
**Status check rather than a claim: three of the four acceptance criteria are already met on `main`, and the fourth needs hardware. Posting so the next person does not rebuild 75% of this.**

I came to take this and checked staleness first. `git log --grep` shows #484 (`06531292`, *"attach the DeepSpeed empty-group guard in every wrapper"*) landed across 22 files, including `tests/test_issue359_deepspeed_guard_coverage.py` — which is named after this issue and does not appear in the thread.

Against the four boxes:

- [x] **Every wrapper via the shared guard, enforced by a test that scans rather than lists.** `_TRAINER_SOURCES = sorted(_TRAINER_DIR.glob("*.py"))` — a directory scan, which is the "a nineteenth wrapper cannot forget it" requirement you wrote. `test_no_deepspeed_capable_module_quietly_loses_the_guard` and `test_the_exemption_list_stays_earned` cover both directions.
- [x] **A control proving the scanner can fail.** `test_the_patterns_would_catch_the_unguarded_shape`, plus `test_a_comment_mentioning_the_guard_does_not_satisfy_it` — which is the sharper one, since a comment is exactly how this would rot.
- [ ] **A non-SFT wrapper verified end-to-end on real multi-GPU hardware.** Not done, and not doable by me: M1, no CUDA. This is the only genuinely open item.
- [x] **Decide and document whether `--deepspeed <file>` is resolved or left as an escape hatch.** Decided *and* documented, in `resolve_user_deepspeed_file` (`utils/deepspeed.py:300`), whose docstring cites this issue: the original path is returned byte-identical when nothing needs changing, so the escape hatch stays one, and a repaired config goes to a temp copy while the user's file is never touched. The rationale is stated too — the two rewritten keys *"are not preferences, they are errors"*.

```
pytest tests/test_issue359_deepspeed_guard_coverage.py    123 passed in 4.0s
grep -rn prune_empty_param_groups src/soup_cli/trainer/   (no hits — the guard
    attaches through utils/deepspeed.py rather than being called per-wrapper,
    which is why a name-based search suggests nothing landed)
```

That last line is why this looked open to me at first, and may be why it looks open to others: searching the trainer package for the guard finds nothing, because #484 attached it centrally rather than wrapper by wrapper. The scanner test is what encodes the coverage.

**The second related gap you listed — ZeRO++ hierarchical partitioning never exercised across two nodes — is also still open** and is the same hardware class as criterion 3.

So: not claiming this. Someone with multi-GPU access could close it by running the `benchmarks/gate-h100-validation.md` STEP 20 template against one non-SFT wrapper. If you would rather this issue be narrowed to just that, or closed with the hardware item tracked separately, that is your call — I would rather flag the state than quietly leave three finished boxes looking unfinished.

**Comment by MakazhanAlpamys** [comment]
Confirmed, and thank you for checking staleness before claiming rather than after. I re-ran your evidence on `main` at `88b006ac`:

```
pytest tests/test_issue359_deepspeed_guard_coverage.py   ->  83 passed in 1.37s
grep -rn prune_empty_param_groups src/soup_cli/trainer/  ->  0 hits
utils/deepspeed.py:300  def resolve_user_deepspeed_file(path, *, gpu_count=None) -> str
```

(I count 83 where you counted 123 — worth noting rather than smoothing over, since it probably means one of us collected a second file. It does not change the conclusion: the suite is green and the directory scan is there.)

Your three checked boxes hold. The point I want to keep is the one in your last line: **a name-based search for `prune_empty_param_groups` under `trainer/` returns nothing**, because the guard attaches through `utils/deepspeed.py` rather than being called per wrapper. That is exactly why this issue reads as open to anyone who greps for it, and it is why #484 never showed up in this thread. I am editing the issue body to say so, so the next person does not spend the same twenty minutes.

Narrowing this issue to the one genuinely open item: **a non-SFT wrapper verified end-to-end on real multi-GPU hardware.** Everything else is done and pinned.

If anyone reading has 2+ CUDA GPUs and wants it: the task is one DPO (or ORPO / KTO) run with LoRA under ZeRO-2 and ZeRO-3, confirming the empty no-decay group does not reach torch's scheduler, with the DeepSpeed and accelerate versions recorded. The unit coverage already proves the guard is attached everywhere; what is missing is the proof that it attaches to something that then *works*, which no CPU can supply.

@Srinivasan8888 — you said M1 and no CUDA, so I am not assigning it to you. Standing down on hardware grounds and publishing the 75% so nobody rebuilds it is worth more here than a speculative PR would have been.

**Comment by Srinivasan8888** [comment]
Two things, and the first is a correction you asked for.

## 83 vs 123: your number was right, mine was wrong

You flagged it as "worth noting rather than smoothing over", so I measured each tree rather than guessing which of us collected a second file. It was me:

```
tree                                   collected
7384134  (main when I posted, 09-10)       83
88b006ac (your run, 09-11)                 83
f491ac5  (main today)                      90
```

At the exact tree I measured, the file collects 83, so my "123 passed" cannot have come from this file alone. I don't have the original invocation, so I can't tell you what else it collected, only that the figure was wrong. Apologies for leaving it for you to catch.

The drift since then fully reconciles: #786 added 5 tests (83 → 88), and one new module under `trainer/` (33 → 34 files) adds one item to each of the two functions parametrized over `_TRAINER_SOURCES` (88 → 90). So the directory scan is picking up new wrappers without anyone listing them, which is the property this issue asked for.

## The assignment

This was assigned to me on 09-14 at 05:57, the same minute as #880 and #886. On 09-11 you said you were not assigning it because of hardware, and that hasn't changed: M1, no CUDA. The one open item, a non-SFT wrapper run end-to-end on 2+ CUDA GPUs under ZeRO-2 and ZeRO-3, is not something I can do or fake a CPU substitute for.

If the assignment was meant for #359, tell me what you had in mind and I'll take it. If it was part of that batch, please unassign me so the issue reads as available to someone with the hardware. I would rather ask than leave it looking claimed by someone who can't close it.

**Comment by MakazhanAlpamys** [comment]
**Unassigned — you were right to ask, and the assignment was mine to explain.** It was part of the 09-14 05:57 batch alongside #880 and #886, and I did not re-read my own 09-11 note saying I was deliberately *not* assigning this one because of hardware. So it was a batch slip, not a change of mind, and leaving it on you would have made the issue read as claimed by someone who cannot close it. `help wanted` is back on.

The open item is unchanged and is genuinely hardware-gated: a non-SFT wrapper run end-to-end on 2+ CUDA GPUs under ZeRO-2 and ZeRO-3. There is no honest CPU substitute — the empty no-decay group only materialises when DeepSpeed actually partitions, so a mocked run would prove that the mock behaves, not that the fix does.

**On the 83 vs 123 correction: thank you, and this is the part I want on the record.** You did not smooth it over or leave it ambiguous — you measured three separate trees, found the error was yours, said so plainly, and then reconciled the drift so the remaining number is explained rather than merely different:

```
7384134  (main, 09-10)   83
88b006ac (my run, 09-11) 83
f491ac5  (main today)    90     = 83 + 5 from #786 + 2 from one new trainer module
```

That last line is the useful half and it goes beyond the correction: the directory scan is picking up new wrappers **without anyone listing them**, which is exactly the property #359 asked for. A hand-maintained list would have silently stopped covering the new module; this one gained two cases on its own.

A wrong number corrected by its author, with the method shown, is worth more to this repo than a right number asserted — I have published three of my own wrong figures this week and had to correct them in the same thread, so I am not saying that as a courtesy.

If someone with multi-GPU CUDA wants this, the remaining work is written up in the issue body and the SFT-side fix in `utils/deepspeed.py` is the template.
